src-cpp/rdkafkacpp.h

```diff
@@ -1365,7 +1366,139 @@ class RD_EXPORT MessageTimestamp {
+        return err_;
+    }
+    
+   private:
+    std::string key_;
+    RdKafka::ErrorCode err_;
+    std::string value_container_;
+    void *operator new(size_t); /* Prevent dynamic allocation */
+  };
+
+  /**
+   * @brief create a new instance of the Headers object
+   */
+  static Headers *create(size_t initial_size = 8, bool free_rd_headers = true);
+
+  /**
+   * @brief create a new instance of the Headers object from a std::vector
+   */
+  static Headers *create(const std::vector<Header> &headers, bool free_rd_headers = true);
+
+  /** 
+   * @brief adds a Header to the end
+   *
+   * @returns An ErrorCode signalling a success or failure to add the Header.
+   */
+  virtual ErrorCode add(const std::string& key, const char* value) = 0;
+
+  /** 
+   * @brief removes all the Headers of a given key
+   *
+   * @returns An ErrorCode signalling a success or failure to remove the Header.
+   */
+  virtual ErrorCode remove(const std::string& key) = 0;
+
+  /** 
+   * @brief gets all of the Headers of a given key
+   *
+   * @returns a std::vector containing all the Headers of the given key.
+   */
+  virtual std::vector<Header> get(const std::string &key) const = 0;
+
+  /** 
+   * @brief gets the last occurrence of a Header of a given key
+   *
+   * @returns the Header if found, otherwise a Header with an ErrorCode 
+   */
+  virtual Header get_last(const std::string& key) const = 0;
+
+  /** 
+   * @brief returns all the Headers of a Message
+   *
+   * @returns a std::vector containing all of the Headers of a message
+   */
+  virtual std::vector<Header> get_all() const = 0;
 
+  /** 
+   * @brief the count of all the Headers
+   *
+   * @returns a size_t count of all the headers
+   */
+  virtual size_t size() const = 0;
+
+  /**
+   * @brief Returns the underlying librdkafka C rd_kafka_headers_t handle.
+   *
+   * @warning Calling the C API on this handle is not recommended and there
+   *          is no official support for it, but for cases where the C++ API
+   *          does not provide the underlying functionality this C handle can be
+   *          used to interact directly with the core librdkafka API.
+   *
+   * @remark The lifetime of the returned pointer can be different than the lifetime
+   *         of the Headers message due to how the producev function in the C API works
+   *         if there is no error then the producev will take care of deallocation
+   *         but if there is an error then it is the responsibility of the calling
+   *         object to deallocate the underlying C implementation if an instance
+   *         of the Headers object is created with free_rd_headers set to `false`
+   *
+   * @remark Include <rdkafka/rdkafka.h> prior to including
+   *         <rdkafka/rdkafkacpp.h>
+   *
+   * @returns \c rd_kafka_headers_t*
+   */
+  virtual struct rd_kafka_headers_s *c_headers() = 0;
+
+  /** 
+   * @brief cleans up the underlying alocated C implementation headers if called
```

Comment:
when would you use this method?

Reply:
It's probably an artifact of some experimentation I was doing, with your comment mentioning detaching the headers from the rkmessage I could completely encapsulate the Headers within in this instance and let the destructor manage the lifetime of the rd_headers

Reply:
Actually come to think of it, I think this is still needed in the case where producev fails and doesn't clean up the C headers and in that case you will have to manually clean them up yourself because producev can't set the internal C headers ptr to NULL so the destructor will fail for a double free